File: ios/AnkiMCAT/Sources/AnkiMCAT/Practice/MCATMetrics.swift
// Copyright: Ankitects Pty Ltd and contributors
// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
//
// Pure Performance/Readiness math for the MCAT Practice tab.
//
// Every formula here is specified authoritatively in
// `.factory/runs/2026-07-02-read-practice-tabs/contracts/data-model.md`
// ("Performance formula" / "Readiness formula" sections) and is a field-for-
// field port of `ts/routes/practice/mcatMetrics.ts` -- do not "simplify" or
// approximate anything here; server, desktop-fallback, and iOS-fallback must
// produce identical numbers (cross-platform parity is a testing-domain
// regression test).
//
// Foundation-only (no networking, no SwiftUI) so
// `ios/AnkiMCAT/Tests/PracticeLogicTests/run.sh` can swiftc-compile+run this
// file standalone, with no Simulator or Xcode project needed.

import Foundation

/// The 4 canonical MCAT categories/sections, fixed order.
enum MCATCategory: String, Codable, CaseIterable {
    case bioBiochem = "bio_biochem"
    case chemPhys = "chem_phys"
    case psychSoc = "psych_soc"
    case cars = "cars"
}

struct PracticeHistoryItem {
    var questionId: String
    var category: MCATCategory
    var correct: Bool
    var difficultyB: Double
}

struct FsrsCategorySummary {
    var category: MCATCategory
    var averageRecall: Double // [0,1]
    var masteredFraction: Double // [0,1]
    var enoughData: Bool
    var gradedReviews: Int
}

struct FsrsSummary {
    var perCategory: [FsrsCategorySummary]
    var overallMeanRecall: Double
}

struct PerformanceCategory {
    var category: MCATCategory
    var p: Double // [0,1]
    var pLow: Double // [0,1] -- 90% CI lower bound (0 when !enoughData)
    var pHigh: Double // [0,1] -- 90% CI upper bound (0 when !enoughData)
    var enoughData: Bool
    var n: Int
}

struct Performance {
    var overall: (p: Double, pLow: Double, pHigh: Double, enoughData: Bool, n: Int)
    var perCategory: [PerformanceCategory]
}

enum Confidence: String {
    case high, medium, low
}

struct Readiness {
    var scorePoint: Int // [472,528]
    var scoreLow: Int
    var scoreHigh: Int
    var confidence: Confidence
    var note: String
    var enoughData: Bool
}

enum MCATMetrics {
    /// Minimum number of records required before a Performance figure is shown.
    private static let minN = 5

    /// z-score for a 90% interval (matches the Memory dashboard's 90% CI).
    private static let z90 = 1.645

    /// MAP ability estimate (Rasch / 1-PL), N(0,1) prior, Newton-Raphson.
    ///
    /// theta_0 = 0
    /// p_i(theta) = 1 / (1 + exp(-(theta - b_i)))
    /// theta_{k+1} = theta_k + [ sum_i(r_i - p_i(theta_k)) - theta_k ]
    ///                          / [ sum_i p_i(theta_k)(1 - p_i(theta_k)) + 1 ]
    ///
    /// Iterate up to 25x or until |delta| < 1e-4; clamp final theta to [-4, 4].
    static func estimateAbility(_ records: [(correct: Bool, difficultyB: Double)]) -> Double {
        var theta = 0.0

        for _ in 0..<25 {
            var numerator = -theta
            var denominator = 1.0

            for (correct, difficultyB) in records {
                let p = 1 / (1 + exp(-(theta - difficultyB)))
                numerator += (correct ? 1.0 : 0.0) - p
                denominator += p * (1 - p)
            }

            let delta = numerator / denominator
            theta += delta

            if abs(delta) < 1e-4 {
                break
            }
        }

        return max(-4, min(4, theta))
    }

    /// p = 1 / (1 + exp(-theta)) -- probability of a correct answer on a new,
    /// average-difficulty (b=0) item.
    private static func probabilityAtZeroDifficulty(_ theta: Double) -> Double {
        1 / (1 + exp(-theta))
    }

    /// Posterior standard error of the MAP ability estimate. The N(0,1) prior
    /// contributes precision 1; each item contributes Fisher information
    /// p_i(1-p_i) at the estimate. SE = 1 / sqrt(precision), so the Performance
    /// range narrows as more questions are answered.
    private static func abilityStdError(_ records: [(correct: Bool, difficultyB: Double)],
                                        _ theta: Double) -> Double {
        var precision = 1.0
        for (_, difficultyB) in records {
            let p = 1 / (1 + exp(-(theta - difficultyB)))
            precision += p * (1 - p)
        }
        return 1 / sqrt(precision)
    }

    /// The point estimate plus a 90% interval for P(correct on a new b=0 item),
    /// obtained by pushing the ability CI (theta +/- z*SE) through the logistic.
    private static func performanceEstimate(_ records: [(correct: Bool, difficultyB: Double)])
        -> (p: Double, pLow: Double, pHigh: Double) {
        let theta = estimateAbility(records)
        let se = abilityStdError(records, theta)
        return (
            probabilityAtZeroDifficulty(theta),
            probabilityAtZeroDifficulty(theta - z90 * se),
            probabilityAtZeroDifficulty(theta + z90 * se)
        )
    }

    /// Computes overall and per-category Performance from the full practice
    /// history. Each of overall/per-category is gated independently by the
    /// `N >= 5` minimum -- e.g. a category with 0 answers always shows
    /// "not enough data" even if the overall history is large.
    static func computePerformance(_ history: [PracticeHistoryItem]) -> Performance {
        let overallN = history.count
        var overall: (p: Double, pLow: Double, pHigh: Double, enoughData: Bool, n: Int)
        if overallN >= minN {
            let est = performanceEstimate(history.map { ($0.correct, $0.difficultyB) })
            overall = (est.p, est.pLow, est.pHigh, true, overallN)
        } else {
            overall = (0, 0, 0, false, overallN)
        }

        let perCategory: [PerformanceCategory] = MCATCategory.allCases.map { category in
            let records = history.filter { $0.category == category }
            let n = records.count
            if n >= minN {
                let est = performanceEstimate(records.map { ($0.correct, $0.difficultyB) })
                return PerformanceCategory(category: category, p: est.p, pLow: est.pLow,
                                            pHigh: est.pHigh, enoughData: true, n: n)
            }
            return PerformanceCategory(category: category, p: 0, pLow: 0, pHigh: 0,
                                        enoughData: false, n: n)
        }

        return Performance(overall: overall, perCategory: perCategory)
    }

    private static func clamp(_ value: Double, _ lo: Double, _ hi: Double) -> Double {
        max(lo, min(hi, value))
    }

    private struct SectionResult {
        var scoreSection: Int // rounded int, [118,132]
        var halfwidth: Double // [1,7] (or exactly 7 in the "no data" case)
        var hasData: Bool // performance.enoughData OR fsrs.enoughData for this section
    }

    private static func computeSection(_ perf: PerformanceCategory,
                                        _ fsrs: FsrsCategorySummary) -> SectionResult {
        let performanceHasData = perf.enoughData
        let fsrsHasData = fsrs.enoughData

        var proficiency: Double
        var nEff: Double

        if !performanceHasData && !fsrsHasData {
            // "No data" section: contributes the mean to the point estimate
            // but the maximum half-width to the interval (data-model.md
            // Step 5).
            proficiency = 0.5
            nEff = 0
        } else {
            let mCat = fsrsHasData
                ? clamp(0.6 * fsrs.averageRecall + 0.4 * fsrs.masteredFraction, 0, 1)
                : 0

            if performanceHasData && fsrsHasData {
                proficiency = 0.5 * perf.p + 0.5 * mCat
            } else if performanceHasData {
                proficiency = perf.p
            } else {
                proficiency = mCat
            }

            nEff = Double(perf.n) + 0.2 * Double(fsrs.gradedReviews)
        }

        let scoreSection = Int((clamp(118 + 14 * proficiency, 118, 132)).rounded())
        let halfwidth = clamp(14 / sqrt(1 + nEff), 1, 7)

        return SectionResult(scoreSection: scoreSection, halfwidth: halfwidth,
                              hasData: performanceHasData || fsrsHasData)
    }

    /// Computes the projected MCAT scaled-score Readiness from Performance and
    /// FsrsSummary, per data-model.md's "Readiness formula" Steps 1-5.
    static func computeReadiness(_ performance: Performance, _ fsrs: FsrsSummary) -> Readiness {
        let sections: [SectionResult] = MCATCategory.allCases.map { category in
            let perf = performance.perCategory.first { $0.category == category }
                ?? PerformanceCategory(category: category, p: 0, pLow: 0, pHigh: 0,
                                        enoughData: false, n: 0)
            let fsrsCat = fsrs.perCategory.first { $0.category == category }
                ?? FsrsCategorySummary(category: category, averageRecall: 0, masteredFraction: 0,
                                        enoughData: false, gradedReviews: 0)
            return computeSection(perf, fsrsCat)
        }

        var scorePoint = 0
        var scoreLowSum = 0.0
        var scoreHighSum = 0.0
        var halfwidthSum = 0.0
        var sectionsWithData = 0

        for section in sections {
            scorePoint += section.scoreSection
            scoreLowSum += clamp(Double(section.scoreSection) - section.halfwidth, 118, 132)
            scoreHighSum += clamp(Double(section.scoreSection) + section.halfwidth, 118, 132)
            halfwidthSum += section.halfwidth
            if section.hasData {
                sectionsWithData += 1
            }
        }

        let avgHalfwidth = halfwidthSum / Double(sections.count)
        let confidence: Confidence
        if avgHalfwidth <= 2 {
            confidence = .high
        } else if avgHalfwidth <= 4 {
            confidence = .medium
        } else {
            confidence = .low
        }

        let enoughData = sectionsWithData >= 2

        return Readiness(
            scorePoint: scorePoint,
            scoreLow: Int(scoreLowSum.rounded()),
            scoreHigh: Int(scoreHighSum.rounded()),
            confidence: confidence,
            note: enoughData
                ? ""
                : "Answer more practice questions or review more cards to see a projected score.",
            enoughData: enoughData
        )
    }
}

File: ts/routes/practice/mcatMetrics.ts
// Copyright: Ankitects Pty Ltd and contributors
// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html

/**
 * Pure Performance/Readiness math for the MCAT Practice tab.
 *
 * Every formula here is specified authoritatively in
 * `.factory/runs/2026-07-02-read-practice-tabs/contracts/data-model.md`
 * ("Performance formula" / "Readiness formula" sections) and must match the
 * server and iOS implementations exactly bit-for-bit-ish (cross-platform
 * parity is a testing-domain regression test) -- do not "simplify" or
 * approximate anything in here.
 *
 * No I/O, no Date.now(), no Svelte imports: this module must be importable
 * and testable in complete isolation (e.g. from a future Vitest suite).
 */

export type Category = "bio_biochem" | "chem_phys" | "psych_soc" | "cars";

/** Canonical, fixed order of the 4 MCAT sections/categories. */
export const CATEGORIES: readonly Category[] = [
    "bio_biochem",
    "chem_phys",
    "psych_soc",
    "cars",
];

export interface PracticeHistoryItem {
    question_id: string;
    category: Category;
    correct: boolean;
    difficulty_b: number;
}

export interface FsrsCategorySummary {
    category: Category;
    average_recall: number; // [0,1]
    mastered_fraction: number; // [0,1]
    enough_data: boolean;
    graded_reviews: number;
}

export interface FsrsSummary {
    per_category: FsrsCategorySummary[];
    overall_mean_recall: number;
}

export interface PerformanceCategory {
    category: Category;
    p: number; // [0,1]
    p_low: number; // [0,1] -- 90% CI lower bound (0 when !enough_data)
    p_high: number; // [0,1] -- 90% CI upper bound (0 when !enough_data)
    enough_data: boolean;
    n: number;
}

export interface Performance {
    overall: { p: number; p_low: number; p_high: number; enough_data: boolean; n: number };
    per_category: PerformanceCategory[];
}

export type Confidence = "high" | "medium" | "low";

export interface Readiness {
    score_point: number; // [472,528]
    score_low: number;
    score_high: number;
    confidence: Confidence;
    note: string;
    enough_data: boolean;
}

/** Minimum number of records required before a Performance figure is shown. */
const MIN_N = 5;

/** z-score for a 90% interval (matches the Memory dashboard's 90% CI). */
const Z_90 = 1.645;

/**
 * MAP ability estimate (Rasch / 1-PL), N(0,1) prior, Newton-Raphson.
 *
 * theta_0 = 0
 * p_i(theta) = 1 / (1 + exp(-(theta - b_i)))
 * theta_{k+1} = theta_k + [ sum_i(r_i - p_i(theta_k)) - theta_k ]
 *                          / [ sum_i p_i(theta_k)(1 - p_i(theta_k)) + 1 ]
 *
 * Iterate up to 25x or until |delta| < 1e-4; clamp final theta to [-4, 4].
 */
export function estimateAbility(
    records: { correct: boolean; difficulty_b: number }[],
): number {
    let theta = 0;

    for (let iter = 0; iter < 25; iter++) {
        let numerator = -theta;
        let denominator = 1;

        for (const { correct, difficulty_b } of records) {
            const p = 1 / (1 + Math.exp(-(theta - difficulty_b)));
            numerator += (correct ? 1 : 0) - p;
            denominator += p * (1 - p);
        }

        const delta = numerator / denominator;
        theta += delta;

        if (Math.abs(delta) < 1e-4) {
            break;
        }
    }

    return Math.max(-4, Math.min(4, theta));
}

/** p = 1 / (1 + exp(-theta)) -- probability of a correct answer on a new, average-difficulty (b=0) item. */
function probabilityAtZeroDifficulty(theta: number): number {
    return 1 / (1 + Math.exp(-theta));
}

/**
 * Posterior standard error of the MAP ability estimate. The N(0,1) prior
 * contributes precision 1; each item contributes Fisher information
 * p_i(1-p_i) at the estimate. SE = 1 / sqrt(precision), which shrinks as more
 * questions are answered -- so the Performance range narrows with evidence.
 */
function abilityStdError(
    records: { correct: boolean; difficulty_b: number }[],
    theta: number,
): number {
    let precision = 1;
    for (const { difficulty_b } of records) {
        const p = 1 / (1 + Math.exp(-(theta - difficulty_b)));
        precision += p * (1 - p);
    }
    return 1 / Math.sqrt(precision);
}

/**
 * The point estimate plus a 90% interval for P(correct on a new b=0 item),
 * obtained by pushing the ability CI (theta +/- z*SE) through the logistic.
 */
function performanceEstimate(
    records: { correct: boolean; difficulty_b: number }[],
): { p: number; p_low: number; p_high: number } {
    const theta = estimateAbility(records);
    const se = abilityStdError(records, theta);
    return {
        p: probabilityAtZeroDifficulty(theta),
        p_low: probabilityAtZeroDifficulty(theta - Z_90 * se),
        p_high: probabilityAtZeroDifficulty(theta + Z_90 * se),
    };
}

/**
 * Computes overall and per-category Performance from the full practice
 * history. Each of overall/per-category is gated independently by the
 * `N >= 5` minimum -- e.g. a category with 0 answers always shows
 * "not enough data" even if the overall history is large.
 */
export function computePerformance(history: PracticeHistoryItem[]): Performance {
    const overallN = history.length;
    let overall: { p: number; p_low: number; p_high: number; enough_data: boolean; n: number };
    if (overallN >= MIN_N) {
        overall = { ...performanceEstimate(history), enough_data: true, n: overallN };
    } else {
        overall = { p: 0, p_low: 0, p_high: 0, enough_data: false, n: overallN };
    }

    const per_category: PerformanceCategory[] = CATEGORIES.map((category) => {
        const records = history.filter((r) => r.category === category);
        const n = records.length;
        if (n >= MIN_N) {
            return { category, ...performanceEstimate(records), enough_data: true, n };
        }
        return { category, p: 0, p_low: 0, p_high: 0, enough_data: false, n };
    });

    return { overall, per_category };
}

function clamp(value: number, min: number, max: number): number {
    return Math.max(min, Math.min(max, value));
}

interface SectionResult {
    scoreSection: number; // rounded int, [118,132]
    halfwidth: number; // [1,7] (or exactly 7 in the "no data" case)
    hasData: boolean; // performance.enough_data OR fsrs.enough_data for this section
}

function computeSection(
    perf: PerformanceCategory,
    fsrs: FsrsCategorySummary,
): SectionResult {
    const performanceHasData = perf.enough_data;
    const fsrsHasData = fsrs.enough_data;

    let proficiency: number;
    let nEff: number;

    if (!performanceHasData && !fsrsHasData) {
        // "No data" section: contributes the mean to the point estimate but
        // the maximum half-width to the interval (data-model.md Step 5).
        proficiency = 0.5;
        nEff = 0;
    } else {
        const mCat = fsrsHasData
            ? clamp(0.6 * fsrs.average_recall + 0.4 * fsrs.mastered_fraction, 0, 1)
            : 0;

        if (performanceHasData && fsrsHasData) {
            proficiency = 0.5 * perf.p + 0.5 * mCat;
        } else if (performanceHasData) {
            proficiency = perf.p;
        } else {
            proficiency = mCat;
        }

        nEff = perf.n + 0.2 * fsrs.graded_reviews;
    }

    const scoreSection = Math.round(clamp(118 + 14 * proficiency, 118, 132));
    const halfwidth = clamp(14 / Math.sqrt(1 + nEff), 1, 7);

    return {
        scoreSection,
        halfwidth,
        hasData: performanceHasData || fsrsHasData,
    };
}

/**
 * Computes the projected MCAT scaled-score Readiness from Performance and
 * FsrsSummary, per data-model.md's "Readiness formula" Steps 1-5.
 */
export function computeReadiness(performance: Performance, fsrs: FsrsSummary): Readiness {
    const sections: SectionResult[] = CATEGORIES.map((category) => {
        const perf = performance.per_category.find((p) => p.category === category) ?? {
            category,
            p: 0,
            p_low: 0,
            p_high: 0,
            enough_data: false,
            n: 0,
        };
        const fsrsCat = fsrs.per_category.find((f) => f.category === category) ?? {
            category,
            average_recall: 0,
            mastered_fraction: 0,
            enough_data: false,
            graded_reviews: 0,
        };
        return computeSection(perf, fsrsCat);
    });

    let scorePoint = 0;
    let scoreLow = 0;
    let scoreHigh = 0;
    let halfwidthSum = 0;
    let sectionsWithData = 0;

    for (const section of sections) {
        scorePoint += section.scoreSection;
        scoreLow += clamp(section.scoreSection - section.halfwidth, 118, 132);
        scoreHigh += clamp(section.scoreSection + section.halfwidth, 118, 132);
        halfwidthSum += section.halfwidth;
        if (section.hasData) {
            sectionsWithData++;
        }
    }

    const avgHalfwidth = halfwidthSum / sections.length;
    let confidence: Confidence;
    if (avgHalfwidth <= 2) {
        confidence = "high";
    } else if (avgHalfwidth <= 4) {
        confidence = "medium";
    } else {
        confidence = "low";
    }

    const enoughData = sectionsWithData >= 2;

    return {
        score_point: scorePoint,
        score_low: Math.round(scoreLow),
        score_high: Math.round(scoreHigh),
        confidence,
        note: enoughData
            ? ""
            : "Answer more practice questions or review more cards to see a projected score.",
        enough_data: enoughData,
    };
}
